### PR TITLE
[jit][wip] TVM Integration

### PR DESCRIPTION
**Pull request by bwasti**
This is a first pass at TVM integration.  It sets up basic infrastructure and maps two operators (add, mul) to their Relay equivalent.

TODO:
Default TVM OFF, with CI to test cloning
Python tests

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -82,3 +82,6 @@
 [submodule "third_party/foxi"]
 	path = third_party/foxi
 	url = https://github.com/houseroad/foxi.git
+[submodule "third_party/tvm"]
```
Comment (soumith) [inline .gitmodules]: can we make the user do extra legwork here and actually clone TVM into the pytorch root directory.
I dont want this to be the default submodule that everyone has to clone (TVM and it's submodules end up being a significant chunk of git clone)

File: CMakeLists.txt [hunk 2]
```diff
@@ -138,6 +138,7 @@ option(USE_SNPE "Use Qualcomm's SNPE library" OFF)
 option(USE_SYSTEM_EIGEN_INSTALL
     "Use system Eigen instead of the one under third_party" OFF)
 option(USE_TENSORRT "Using Nvidia TensorRT library" OFF)
+option(USE_TVM "Using TVM library" ON)
```
Comment (soumith) [inline CMakeLists.txt]: by default it's OFF, because of the above comment

Comment (soumith) [reply CMakeLists.txt]: also let's cover the ON case in CI soon

**Review by soumith — CHANGES_REQUESTED** [review]

File: .gitmodules (hunk 1 above)
Comment (apaszke) [reply .gitmodules]: Also, why are we using Facebook's fork?

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 3]
```diff
@@ -1341,5 +1360,11 @@ void FuseGraph(std::shared_ptr<Graph>& graph) {
   }
 }
 
+void CustomFuseGraph(std::shared_ptr<Graph>& graph, std::function<bool(Node*)> fn, Symbol tag) {
+  if (canFuseOnCPU() || canFuseOnGPU()) {
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: Why do we even check this if it has no connection to what our builtin fuser can handle?

Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Also, `tag` is called a `kind` everywhere in our codebase.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 4]
```diff
@@ -187,9 +193,22 @@ struct GraphFuser {
   }
 
   bool isFusable(Node* node) {
+    if (callback_) {
+      return isFusableCallback(node);
+    }
     return isFusableMap(node) || isFusableBatchNorm(node);
   }
 
+  bool isFusableCallback(Node* node) {
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: This function seems unnecessary? Just inline its body into `isFusable`

**Review by apaszke — CHANGES_REQUESTED** [review]
If we're not going to include TVM as a submodule by default, I'd really like us to avoid putting all this code into JIT core. It would be much nicer to design an interface that lets people plug in their own codegens into the JIT, because TVM is really not the only one. Right now we kind of hackily hijack the fuser infrastructure which is less than ideal.

File: .gitmodules (hunk 1 above)
Comment (bwasti) [reply .gitmodules]: @apaszke it's got a patch that enables the C++ frontend (hasn't been landed into dmlc yet)

**Comment by bwasti** [comment]
@soumith should we instead think about fixing up third_party (which seems to be a dumping ground for libraries), and only pulling down necessary libs?  fwiw, TVM only pulls 4 submodules, all of which are relatively small

```
(python3_dev) devvm304.prn3.facebook.com:/home/bwasti/local $ time git clone https://github.com/dmlc/tvm.git
Cloning into 'tvm'...
remote: Enumerating objects: 49, done.
remote: Counting objects: 100% (49/49), done.
remote: Compressing objects: 100% (36/36), done.
remote: Total 35127 (delta 13), reused 29 (delta 13), pack-reused 35078
Receiving objects: 100% (35127/35127), 11.60 MiB | 19.32 MiB/s, done.
Resolving deltas: 100% (22822/22822), done.

real    0m1.738s
user    0m1.161s
sys     0m0.242s
(python3_dev) devvm304.prn3.facebook.com:/home/bwasti/local $ cd tvm
(python3_dev) devvm304.prn3.facebook.com:/home/bwasti/local/tvm $ time git submodule update --init --recursive
Submodule 'HalideIR' (https://github.com/dmlc/HalideIR) registered for path '3rdparty/HalideIR'
Submodule 'dlpack' (https://github.com/dmlc/dlpack) registered for path '3rdparty/dlpack'
Submodule 'dmlc-core' (https://github.com/dmlc/dmlc-core) registered for path '3rdparty/dmlc-core'
Submodule '3rdparty/rang' (https://github.com/agauniyal/rang) registered for path '3rdparty/rang'
Cloning into '/data/users/bwasti/tvm/3rdparty/HalideIR'...
Cloning into '/data/users/bwasti/tvm/3rdparty/dlpack'...
Cloning into '/data/users/bwasti/tvm/3rdparty/dmlc-core'...
Cloning into '/data/users/bwasti/tvm/3rdparty/rang'...
Submodule path '3rdparty/HalideIR': checked out '86351c40824dfc4cbb7447d70e5e63d9bd76eb90'
Submodule path '3rdparty/dlpack': checked out '5c792cef3aee54ad8b7000111c9dc1797f327b59'
Submodule path '3rdparty/dmlc-core': checked out '2b5b1ba9c1103f438d164aca32da7cffd8cd48e8'
Submodule path '3rdparty/rang': checked out 'cabe04d6d6b05356fa8f9741704924788f0dd762'

real    0m1.671s
user    0m0.402s
sys     0m0.271s
(python3_dev) devvm304.prn3.facebook.com:/home/bwasti/local/tvm $ du -h . -d 0
35M     .
```

@apaszke The interface is trivial: supply a function to CustomGraphFuse and then register something that can run the fused op.  We can incrementally expose things like threadpools and allocators so that the registered op can run more seamlessly with the rest of pytorch.

I can move `tvm.*` (which exposes the graph fusion function and does a static registration of the operator) to a new folder -- where should I put it? `torch/csrc/jit/backends`, `torch/csrc/backends` (like the python backends folder), `torch/csrc/tvm` (like cuda)?

**Comment by ZolotukhinM** [comment]
> If we're not going to include TVM as a submodule by default, I'd really like us to avoid putting all this code into JIT core. It would be much nicer to design an interface that lets people plug in their own codegens into the JIT, because TVM is really not the only one. Right now we kind of hackily hijack the fuser infrastructure which is less than ideal.

I agree with this. I think that TVM should be used as the first example in establishing a general scheme of how any backend could be integrated. As part of this scheme we need to figure out:
1) Where do the backend sources live?
2) How they are built, what's expected from the build (e.g. an .so with a function with a specific signature)?
3) How is a backend registered in pytorch? I.e. how do we let pytorch know that the backend exists.
4) What is the way of feeding subgraphs into the backend and getting a compiled code for them from it?

These all are general questions, not specific to TVM. This PR mostly covers (4) and plows through (1)-(3), but I'd really like to see a complete story that any other backend could also use as a result of this work.

**Comment by soumith** [comment]
The reasonable and ideal story is:

- the pluggable backend actually lives out of source, like https://github.com/pytorch/xla
- is available as a separate package, maybe like `pip install torch-tvm`.
- registers itself explicitly by the user importing it. for example: `import torch_tvm` will then invoke relevant JIT API to register itself

**Comment by soumith** [comment]
> should we instead think about fixing up third_party (which seems to be a dumping ground for libraries), and only pulling down necessary libs? fwiw, TVM only pulls 4 submodules, all of which are relatively small

@bwasti yes, there is no reasonable answer at the moment the packages in there are grandfathered in from caffe2 and pytorch. I can see a plausible world where we delete all submodules except ones that are in the default build, and the rest are runtime errors (if you try to build with the option) that tell you what additional command to run

File: test/cpp/jit/test_misc.h [hunk 5]
```diff
@@ -347,6 +349,86 @@ void testATenNativeBatchNorm() {
...
+  auto a = SymbolicVariable::asNewInput(*graph, type);
+  auto b = SymbolicVariable::asNewInput(*graph, type);
+  auto c = SymbolicVariable::asNewInput(*graph, type);
+  auto d = a * b;
+  auto e = c * d;
+  graph->registerOutput(e.value());
+  AT_ASSERT(a.value()->isTensor());
+
+  torch::jit::overrideCanFuseOnCPU(true);
+  CustomFuseGraph(graph, torch::jit::relay::isSupported, prim::TVMGroup);
```
Comment (yinghai) [inline test/cpp/jit/test_misc.h]: Can we reuse this code for generic backend lowering?

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 6]
```diff
@@ -187,9 +193,22 @@ struct GraphFuser {
   }
 
   bool isFusable(Node* node) {
+    if (callback_) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/graph_fuser.cpp]: It is weird to only partially use isFusableCallback. The default behavior in the graph fuser should itself be turned into a callback and then we should use that callback for prim::FusionGroup creation.

File: aten/src/ATen/core/interned_strings.h [hunk 7]
```diff
@@ -28,6 +28,7 @@ namespace c10 {
   _(prim, Eval)                    \
   _(prim, Expand) /* onnx */       \
   _(prim, FusionGroup)             \
+  _(prim, TVMGroup)             \
```
Comment (zdevito) [inline aten/src/ATen/core/interned_strings.h]: There is not need to add backend-specific entries to interned_strings.  `Symbol::fromQualString("tvm::TVMGroup")` will work fine. Code generators should not add ops to the `prim` namespace. That is reserved for core primitives.

**Review by zdevito — COMMENTED** [review]
I have minor comments on the fusion pass changes. The TVM-specific parts of build need to be arranged to be in a separate compilation unit similar to the XLA build.

**Pull request closed without merging**